python/lsst/pipe/tasks/processImage.py

```diff
@@ -132,101 +111,67 @@ def makeIdFactory(self, sensorRef):
+
     @pipeBase.timeMethod
-    def process(self, dataRef, inputExposure, enableWriteSources=True):
+    def process(self, dataRef, inputExposure, idFactory=None, backgrounds=None, enableWriteSources=True):
         """Process an Image
-        
+
         @param dataRef: data reference that corresponds to the input image
-        @param inputExposure:  exposure to process
+        @param inputExposure: exposure to process
+        @param idFactory: afw.table.IdFactory to use for source catalog
+        @param backgrounds: afwMath.BackgroundList to be appended to
         @param enableWriteSources: if True then writing sources is allowed.
             Set False if you need to postprocess sources before writing them.
 
         @return pipe_base Struct containing these fields:
-        - postIsrExposure: exposure after ISR performed if calib.doIsr or config.doCalibrate, else None
-        - exposure: calibrated exposure (calexp): as computed if config.doCalibrate,
-            else as upersisted and updated if config.doDetection, else None
-        - calib: object returned by calibration process if config.doCalibrate, else None
+        - exposure: input exposure (as modified in the course of processing)
         - sources: detected source if config.doPhotometry, else None
+        - matches: matches between detected sources and astrometric reference catalog
+        - matchMeta: metadata for matches
+        - backgrounds: background list (input as modified)
         """
-        idFactory = self.makeIdFactory(dataRef)
-
-        # initialize outputs
-        calExposure = inputExposure
-        calib = None
-        sources = None
-        backgrounds = afwMath.BackgroundList()
-        if self.config.doCalibrate:
-            calib = self.calibrate.run(inputExposure, idFactory=idFactory)
-            calExposure = calib.exposure
-            if self.config.doWriteCalibrate:
-                dataRef.put(calib.sources, self.dataPrefix + "icSrc")
-                if calib.matches is not None and self.config.doWriteCalibrateMatches:
-                    normalizedMatches = afwTable.packMatches(calib.matches)
-                    normalizedMatches.table.setMetadata(calib.matchMeta)
-                    dataRef.put(normalizedMatches, self.dataPrefix + "icMatch")
-            try:
-                for bg in calib.backgrounds:
-                    backgrounds.append(bg)
-            except TypeError:     
-                backgrounds.append(calib.backgrounds)
-            except AttributeError:
-                self.log.warn("The calibration task did not return any backgrounds. " +
-                    "Any background subtracted in the calibration process cannot be persisted.")
-        else:
-            calib = None
+        if backgrounds is None:
+            backgrounds = afwMath.BackgroundList()
 
         if self.config.doDetection:
+            if idFactory is None:
+                idFactory = self.makeIdFactory(dataRef)
```

Comment:
No.  I added this as the introduction of idFactory in processCcd.py meant that it was being computed once (I discovered this when the demo failed due to mismatched object IDs).  Here, there actually is a case where the idFactory is being passed in (and I thought I should still allow for the case that it may not be from another caller).